source-dropbox/main.go

```diff
@@ -0,0 +1,300 @@
+}
+
+type advancedConfig struct {
+	AscendingKeys bool `json:"ascendingKeys,omitempty"`
+}
+
+func (c config) Validate() error {
+	requiredProperties := [][]string{
+		{"Path", c.Path},
+	}
+	for _, req := range requiredProperties {
+		if req[1] == "" {
+			return fmt.Errorf("missing '%s'", req[0])
+		}
+	}
+	return nil
+}
+
+func (c config) DiscoverRoot() string {
+	return c.Path
+}
+
+func (c config) RecommendedName() string {
+	return c.Path[1:]
+}
+
+func (c config) FilesAreMonotonic() bool {
+	return c.Advanced.AscendingKeys
+}
+
+func (c config) ParserConfig() *parser.Config {
+	return c.Parser
+}
+
+func (c config) PathRegex() string {
+	return c.MatchKeys
+}
+
+func newDropboxStore(ctx context.Context, cfg config) (*dropboxStore, error) {
+	oauthClient := cfg.Credentials.GetOauthClient(ctx)
+	config := dropbox.Config{
+		Client:   oauthClient,
+		LogLevel: dropbox.LogDebug,
+	}
+	client := files.New(config)
+
+	store := dropboxStore{
+		client: client,
+		config: cfg,
+	}
+
+	if err := store.check(); err != nil {
+		return &dropboxStore{}, err
+	}
+
+	return &store, nil
+}
+
+type dropboxStore struct {
+	client files.Client
+	config config
+}
+
+// check checks if the specified path exists in the Dropbox store.
+// It returns an error if the path does not exist or if there was an error
+// while checking the path.
+func (dbStore *dropboxStore) check() error {
+	_, err := dbStore.client.ListFolder(&files.ListFolderArg{Path: dbStore.config.Path, Limit: 1})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (dbStore *dropboxStore) List(ctx context.Context, query filesource.Query) (filesource.Listing, error) {
+	// Get all files at the specified path from Dropbox.
+	arg := &files.ListFolderArg{Path: query.Prefix, Recursive: query.Recursive}
+	f, err := dbStore.client.ListFolder(arg)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving folder: %w", err)
+	}
+
+	entries := f.Entries
+
+	for f.HasMore {
+		arg := files.NewListFolderContinueArg(f.Cursor)
+
+		f, err = dbStore.client.ListFolderContinue(arg)
+		if err != nil {
+			return nil, fmt.Errorf("retrieving additional files: %w", err)
+		}
+
+		entries = append(entries, f.Entries...)
+	}
+
+	// Convert each file and folder entry to an ObjectInfo.
+	var objects []filesource.ObjectInfo
+	for _, entry := range entries {
+		switch e := entry.(type) {
+		case *files.FileMetadata:
+			obj := filesource.ObjectInfo{
+				Path:       e.PathLower,
+				IsPrefix:   false,
+				ContentSum: e.ContentHash,
+				Size:       int64(e.Size),
+				ModTime:    e.ServerModified,
+			}
+			objects = append(objects, obj)
+
+		case *files.FolderMetadata:
+			if !query.Recursive {
+				obj := filesource.ObjectInfo{
+					Path:     e.PathLower,
+					IsPrefix: true,
+				}
+				objects = append(objects, obj)
+			}
+
+		default:
+			// Ignore all other metadata types (like DeletedMetadata).
+			continue
+		}
+	}
+
+	// Remove files that are lexicographically less than the query's StartAt property.
```

Comment:
We could do the conversion and filtering in a single pass, something like this:

```go
	for _, entry := range entries {
		var obj filesource.ObjectInfo

		switch e := entry.(type) {
		case *files.FileMetadata:
			obj = filesource.ObjectInfo{
				Path:       e.PathLower,
				IsPrefix:   false,
				ContentSum: e.ContentHash,
				Size:       int64(e.Size),
				ModTime:    e.ServerModified,
			}

		case *files.FolderMetadata:
			if query.Recursive {
				continue
			}

			obj = filesource.ObjectInfo{
				Path:     e.PathLower,
				IsPrefix: true,
			}

		default:
			// Ignore all other metadata types (like DeletedMetadata).
			continue
		}

		if query.StartAt < obj.Path {
			// Remove files that are lexicographically less than the query's StartAt property.
			objects = append(objects, obj)
		}
	}
```

Reply:
Agreed, I wasn't sure if combining them would make it a little harder to follow the code. But looking at it, what's happening is still pretty clear. I'll make that change.